internal/artifacts/songofdayspast/songofdayspast.go

```diff
@@ -0,0 +1,119 @@
+package songofdayspast
+
+import (
+	"fmt"
+
+	"github.com/genshinsim/gcsim/pkg/core"
+	"github.com/genshinsim/gcsim/pkg/core/attacks"
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/event"
+	"github.com/genshinsim/gcsim/pkg/core/info"
+	"github.com/genshinsim/gcsim/pkg/core/keys"
+	"github.com/genshinsim/gcsim/pkg/core/player"
+	"github.com/genshinsim/gcsim/pkg/core/player/character"
+	"github.com/genshinsim/gcsim/pkg/modifier"
+)
+
+const (
+	yearningKey        = "yearning-effect"
+	wavesOfDaysPastKey = "waves-of-days-past"
+)
+
+func init() {
+	core.RegisterSetFunc(keys.SongOfDaysPast, NewSet)
+}
+
+type Set struct {
+	Index      int
+	core       *core.Core
+	char       *character.CharWrapper
+	healStacks float64
+}
+
+func (s *Set) SetIndex(idx int) { s.Index = idx }
+func (s *Set) Init() error      { return nil }
+
+func NewSet(c *core.Core, char *character.CharWrapper, count int, param map[string]int) (info.Set, error) {
+	s := Set{
+		core:       c,
+		char:       char,
+		healStacks: 0,
+	}
+
+	if count >= 2 {
+		m := make([]float64, attributes.EndStatType)
+		m[attributes.Heal] = 0.15
+		char.AddStatMod(character.StatMod{
+			Base:         modifier.NewBase("sodp-2pc", -1),
+			AffectedStat: attributes.Heal,
+			Amount: func() ([]float64, bool) {
+				return m, true
+			},
+		})
+	}
+
+	if count >= 4 {
+		c.Events.Subscribe(event.OnHeal, OnHeal(&s), fmt.Sprintf("sodp-4pc-heal-accumulation-%v", char.Base.Key.String()))
+		c.Events.Subscribe(event.OnEnemyHit, OnEnemyHit(&s), fmt.Sprintf("waves-of-days-past-%v", char.Base.Key.String()))
+	}
+
+	return &s, nil
+}
+
+func OnHeal(s *Set) func(args ...interface{}) bool {
+	return func(args ...interface{}) bool {
+		if s.core.Status.Duration(wavesOfDaysPastKey) > 0 {
+			return false
+		}
+		src := args[0].(*player.HealInfo)
+		healAmt := args[2].(float64)
+		if src.Caller != s.char.Index {
+			return false
+		}
+		s.healStacks += healAmt
+		if s.healStacks >= 15000 {
+			s.healStacks = 15000
+		}
+		if s.core.Status.Duration(yearningKey) == 0 {
+			s.core.Status.Add(yearningKey, 6*60)
+			s.core.Tasks.Add(func() {
+				s.core.Status.Add(wavesOfDaysPastKey, 10*60)
+				s.core.Flags.Custom[wavesOfDaysPastKey] = 5
```

Comment:
since another yearning instance can already start while waves effect is still up, the healStacks/collected healing needs to "snapshot" into another set var and then set healStacks to 0 in here